### PR TITLE
[TH2-4929] Transport protocol

### PR DESCRIPTION
**Pull request by lumber1000**

### CODE REVIEW

File: src/main/kotlin/com/exactpro/th2/kafka/client/KafkaConnection.kt [hunk 1]
```diff
@@ -40,50 +46,49 @@ import org.apache.kafka.common.errors.TimeoutException
...
-        val alias = message.sessionAlias ?: error("Message '${message.id.logId}' does not contain session alias.")
-        val kafkaStream = config.aliasToTopicAndKey[alias] ?: KafkaStream(config.aliasToTopic[alias]?.topic ?: error("Session alias '$alias' not found."), null)
-        val value = message.body.toByteArray()
-        val messageIdBuilder = message.id.toBuilder().apply {
-            direction = Direction.SECOND
-            bookName = factory.boxConfiguration.bookName
-            setConnectionId(connectionIdBuilder.setSessionGroup(config.aliasToSessionGroup.getValue(alias)))
-        }
+    protected abstract val MESSAGE.logId: String
+    protected abstract val MESSAGE.messageSessionAlias: String?
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/kafka/client/KafkaConnection.kt]: Maybe change type to not nullable and encapsulate error into getter

File: src/main/kotlin/com/exactpro/th2/kafka/client/Main.kt [hunk 2]
```diff
@@ -63,53 +66,98 @@ fun main(args: Array<String>) {
...
+                    runCatching {
+                        connection.publish(message)
+                    }.onFailure {
+                        val errorText = "Could not publish message ${message.id.logId}. Consumer tag ${metadata.consumerTag}"
+                        LOGGER.error(it) { errorText }
+                        eventSender.onEvent(errorText, "SendError", message.id.toProto(batch), it)
+                    }
+                }
+            }
+            { transportRouter.subscribeAll(transportListener, INPUT_QUEUE_ATTRIBUTE) }
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/kafka/client/Main.kt]: Use different set of attributes for transport and protobuf

File: src/main/kotlin/com/exactpro/th2/kafka/client/RawMessageProcessor.kt [hunk 3]
```diff
@@ -1,14 +1,15 @@
 package com.exactpro.th2.kafka.client
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/kafka/client/RawMessageProcessor.kt:17]: Copyright

File: src/main/kotlin/com/exactpro/th2/kafka/client/Utils.kt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.exactpro.th2.kafka.client
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/kafka/client/Utils.kt]: Copyright

File: src/main/kotlin/com/exactpro/th2/kafka/client/Utils.kt [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+
+import com.exactpro.th2.common.schema.message.impl.rabbitmq.transport.MessageId
+
+val MessageId.logId: String // TODO: move to common utils?
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/kafka/client/Utils.kt]: `TODO: move to common utils?` - yes

File: src/main/kotlin/com/exactpro/th2/kafka/client/Main.kt (hunk 2 above)
Comment (lumber1000) [reply src/main/kotlin/com/exactpro/th2/kafka/client/Main.kt]: Attributes are different: transport router adds "transport-group" during subscribe and raw batch router adds "raw" attribute.

Comment (Nikita-Smirnov-Exactpro) [reply src/main/kotlin/com/exactpro/th2/kafka/client/Main.kt]: O, we use deprecated messageRouterRawBatch which add raw attribute. OK

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

File: src/main/kotlin/com/exactpro/th2/kafka/client/Config.kt [hunk 6]
```diff
@@ -194,7 +194,9 @@ class Config(
     val createTopics: Boolean = false,
     val topicsToCreate: List<String> = emptyList(),
     val newTopicsPartitions: Int = 1,
-    val newTopicsReplicationFactor: Short = 1
+    val newTopicsReplicationFactor: Short = 1,
+
+    val useTransport: Boolean = false
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/kafka/client/Config.kt:202]: Please add information into readme

**Review by Nikita-Smirnov-Exactpro — CHANGES_REQUESTED** [review]

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

File: src/main/kotlin/com/exactpro/th2/kafka/client/KafkaConnection.kt [hunk 7]
```diff
@@ -224,6 +234,28 @@ abstract class KafkaConnection<MESSAGE, MESSAGE_BUILDER>(
 
     override fun close() = producer.close()
 
+    private fun createTopicOffsets(topic: String) =
+        TopicOffsets { partition, prevOffset, newOffset ->
+            eventSender.onEvent(
+                name = "Gap detected for topic '$topic' in partition '$partition' between $prevOffset and $newOffset",
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/kafka/client/KafkaConnection.kt:240]: Maybe add information about subscription group

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

File: src/main/kotlin/com/exactpro/th2/kafka/client/KafkaConnection.kt (hunk 7 above)
Comment (OptimumCode) [reply src/main/kotlin/com/exactpro/th2/kafka/client/KafkaConnection.kt:240]: @Nikita-Smirnov-Exactpro Do you mean `consumer group`?

**Pull request merged**